Expose capture suffixes in queries

hendrikvanantwerpen opened the pull request:
This PR adds support for exposing the pattern suffixes like `?` and `*` as part of the query object. This is useful information for downstream tools that are consuming a query match. In my use case, an optional pattern should map to an `Option` type, while a list or star pattern should map to `Vec` type.

## Implementation notes

The suffixes are exposed in a separate array `capture_suffixes` that has the same shape as the `capture_names` array. This way, the change should not be breaking for users.

~The C code exposes the suffixes as characters, but in the Rust bindings I convert them to a more semantic enum.~

maxbrunsfeld commented:
### Test results

> I have tried to verify the tests by running ./script/generate-fixtures followed by ./script/test. Unfortunately, there were many failing tests.

Can you share the failure messages? I can't yet tell what might be happening on your machine based on what you've shared. The CI seems to be passing though.

### Implementation  Notes

> The C code exposes the suffixes as characters

I don't think this approach is going to allow you to pick the correct data type for a given capture in the general. For example, take a pattern like this:

```clj
(class_declaration
  name: (type_identifier) @class_name
  (class_heritage
    (extends_clause
      (type_identifier) @superclass_name))?
)
```

Here, the `superclass_name` is optional, but the `?` operator is not applied directly to the `type_identifier` node that has the `superclass_name` capture. Rather, it's applied two levels higher up the pattern tree.

Similarly, a capture may be optional even if the `?` operator isn't used at all, as in this pattern:

```clj
(call_expression
  function: (identifier) @function-name
  arguments: (argument_list
    [
      (integer_literal) @integer_argument
      (string_literal) @string_argument
    ]))
```

Because of the alternative expression (delimited by square brackets) the matches for this pattern will contain *either* an `integer_argument` or an `string_literal`.

Finally, a capture can appear multiple times within a pattern even without the use of a repetition operator:

```clj
(tuple
  (_) @tuple_member
  (_) @tuple_member)
```

So I don't think that exposing an operator suffix is the right level of abstraction for characterizing a given capture.

### Alternative API

I think that to properly characterize the data type needed to represent each capture, we'll need to do something that's less coupled to the surface syntax of the query, and based on a slightly deeper inspection of the state machine that we're building.

I think the API needs to take a `pattern_index`, because captures can be used differently in different patterns within a query. And I'd prefer to phrase it without making reference to the operator characters, for the reasons mentioned above. Maybe it could just return two separate boolean (as out-parameters) that indicate if the capture is *optional*, and if it can appear *multiple* times.

```c
void ts_query_quantity_for_capture(
  const TSQuery *query,
  uint32_t pattern_index,
  uint32_t capture_index,
  bool *is_required,
  bool *is_multiple
);
```

### Alternative Implementation

I'm not totally sure how to implement this API. It's possible that you could do it at parse time, while building the state machine. But it might be easier to do it in a separate pass, at the end of `ts_query_new`, once the state machine is already built, but walking the state machine for each pattern, and recording how many times each capture occurs.

We already do something like this (but much more complex) for determining which steps are "fallible", and which steps are guaranteed to eventually match. That logic lives in `ts_query__analyze_patterns`. I think that this new analysis should be completely separate from that though, because unlike `ts_query__analyze_patterns`, this logic doesn't need to make reference to the parse table (and abstractly interpret every possible parse tree). It depends only on the state machine itself.

hendrikvanantwerpen commented:
Thanks for the detailed feedback @maxbrunsfeld!

If I understand you correctly, the following are at least the case:

- Captures are subject to the suffixes of all enclosing patterns.
- If captures appear multiple times it (may) result in multiple values for that capture.

I'm also wondering, although it's probably irrelevant for this problem, if the following is the case:

- Is only one alternative chosen from an alternative list `[ ... ]`, or could multiple entries match (and this capture)?

I will take your suggested API, and see if I can implement that. Instead of having two booleans, we could perhaps use an enum representing `1`, `+`, `*`, and `?`. The value for a capture occurrence would be the join (I think?) of all suffixes that apply (i.e., all suffixes of the captured pattern and enclosing patterns):

    1 \/ q = q

    ? \/ 1 = ?
    ? \/ + = *
    ? \/ q = q

    + \/ ? = *
    + \/ * = *
    + \/ q = q

    * \/ _ = *

On my side I would probably combine captures with the same name on different patterns in the same way, but that doesn't matter much for this API.

maxbrunsfeld commented:
For unit-testing this new API, I would recommend creating some kind of table-driven test, similar to [this one](https://github.com/tree-sitter/tree-sitter/blob/master/cli/src/tests/query_test.rs#L3516-L3532) that can check the quantifier results for many different patterns.

hendrikvanantwerpen commented:
Thanks for the pointer to the tests. I think the approach I have implemented now works well and gives reasonably precise results. I'll whip up some tests next week for the full picture.

hendrikvanantwerpen commented:
In the end I stayed with the interface I originally proposed, exposing merged quantifiers corresponding to captures. The reason is that this allows more precise quantifiers, that are hard to compute from the API you proposed. For example, if all branches of an alternation have the same capture with the same quantifier, the quantifier for that capture can be kept the same overall as it is in each branch. If we know the quantifier for each occurrence of the capture, we also need to know the structure of the query to reconstruct this kind of merged result. This structure is currently not exposed. Also, the values resulting from captures are also merged, so there is not much use for quantifiers per capture if we do not know which values correspond to them.

I have added a bunch of tests in the format you suggested to show which quantifiers are computed for different scenarios.

hendrikvanantwerpen commented:
CI fails at `Install emscripten` with `Error: The operation was canceled.`. Not sure if that is related to my changes.

Locally I am able to run all tests successfully at least (previously I missed a step for fixture generation).

maxbrunsfeld commented:
I think one build is getting cancelled because the other build failed w/ C compile errors.

hendrikvanantwerpen commented:
> I think one build is getting cancelled because the other build failed w/ C compile errors.

You are right. Some errors are not generated by my local `make` invocation, unfortunately.

maxbrunsfeld commented:
> If we know the quantifier for each occurrence of the capture, we also need to know the structure of the query to reconstruct this kind of merged result. This structure is currently not exposed.

I think the issue I'm talking about is a bit simpler than this ☝️. A query consists of many top-level patterns. And whenever a query matches, it indicates which *pattern* it was that matched. One of the main benefits of the query engine (compared to arbitrary tree traversals) is that you can use a single query to efficiently search for arbitrarily many patterns in a single pass, and then as you find matches, you can apply different handling to each pattern. So when characterizing a query, it makes sense to specify which pattern you're talking about.

For example, consider this query:

```clj
(function_definition name: (identifier) @x)

(type_identifier) @y
(property_identifier) @z

; ... many other patterns involving `@x`

(array (identifier)* @x)
```

What is the quantifier for the capture `@x`? When the first pattern matches, `@x` will always appear exactly once. When the last pattern matches, `@x` can appear zero-or-more times. We don't need to merge these distinct values into one global quantity; there should just be a quantity per pattern.

If that still seems wrong to you, let me know what your concerns are, and how it affects your use-case.

hendrikvanantwerpen commented:
It looks like I misunderstood you. What you are describing is an optimization that I'd indeed want to use. You are right that in that case the quantifiers should not be merged.

I have changed the API to match your proposal, and added a test for queries with multiple patterns as well.

Also let me know if the naming of things follows existing nomenclature, or if other names make more sense.

maxbrunsfeld commented:
Ok, your code is looking very good, and In particular, the tests you wrote look great.

I haven't given this a careful read yet, and I may need a few days before I have the focused time. Thanks so much for all the work you've put into this.

hendrikvanantwerpen commented:
/cc @dcreager for review

maxbrunsfeld commented on lib/src/query.c:
```diff
@@ -1975,14 +2311,22 @@ static TSQueryError ts_query__parse_pattern(
+    if (e) {
+      capture_quantifiers_delete(&field_capture_quantifiers);
+      return e;
+    }
```
Maybe now it'd be cleaner to unify these two conditionals:

```c
if (e) {
  capture_quantifiers_delete(&field_capture_quantifiers);
  if (e == PARENT_DONE) e = TSQueryErrorSyntax;
  return e;
}
```

maxbrunsfeld reviewed (commented):
This is looking really good. I left a bunch of fairly minor suggestions.

maxbrunsfeld commented on lib/src/query.c:
```diff
@@ -119,6 +119,11 @@ typedef struct {
+/**
+ * CaptureQuantififers - a data structure holding the quantifiers of pattern captures.
+ */
+typedef Array(TSQuantifier) CaptureQuantifiers;
```
Optimization thoughts -

Currently, we allocate and free one of these lists on each sub-pattern of a query that contains a capture anywhere inside of it. And the length of the list is proportional to the numeric value of the largest capture id, so if a query has 50 captures, we'll allocate a 50-element list for any sub pattern *containing* the 50th capture.

For that reason, I wonder if it's worth making this `Array(uint8_t)`, and casting to `TSQuantifier` when reading from the array, so that we'll only use one byte per member. Currently, I think it'll represent `TSQuantifier` as an `int` (so 4 bytes).

It's a micro-optimization, but a pretty easy one to do. Thoughts?

hendrikvanantwerpen replied:
That seems like it might be useful, and will be contained to the implementations of the `capture_quantifier*` procedures. I'll implement that change.

hendrikvanantwerpen commented on lib/src/query.c:
```diff
@@ -452,6 +458,296 @@ static void capture_list_pool_release(CaptureListPool *self, uint16_t id) {
+  switch (left)
+  {
+    case Zero:
+      result = Zero;
```
That is what I had at first, but the compiler kept complaining about a missing default case. Since there is no meaningful default, I wrote it like this, so the uninitialized state acts as a default. This default is never returned (bar invalid inputs) since the case is exhaustive. However, the compiler (in CI at least) wouldn't realize that 🤷🏻.

If you prefer early returns, I can write it as such, use an arbitrary default value and add a big comment explaining that the default value is really meaningless.

hendrikvanantwerpen replied:
Actually, after rereading the code, I see that the result is not even uninitialized. I'll compact this a bit, then :+1:

hendrikvanantwerpen commented on lib/src/query.c:
```diff
@@ -1711,12 +2013,20 @@ static TSQueryError ts_query__parse_pattern(
         return e;
       }
 
+      if(start_index == 0) {
```
Correct, my mental model was incorrect here. I've also added a test covering this case.

hendrikvanantwerpen commented on lib/src/query.c:
```diff
@@ -1339,7 +1635,7 @@ static bool ts_query__analyze_patterns(TSQuery *self, unsigned *error_offset) {
 
   // Mark as indefinite any step with captures that are used in predicates.
-  Array(uint16_t) predicate_capture_ids = array_new();
+  Array(uint16_t) predicate_capture_ids = array_new(); // FIXME
```
No, I'll drop them.

hendrikvanantwerpen commented on lib/src/query.c:
```diff
@@ -2082,6 +2436,10 @@ static TSQueryError ts_query__parse_pattern(
+        // Add only once, not for every branch, lest the quantifier will be '+' instead of '1'
+        if (step_index == starting_step_index) {
+          capture_quantifiers_add_for_id(capture_quantifiers, capture_id, One);
+        }
```
Yes! I'll do that.

hendrikvanantwerpen commented:
@maxbrunsfeld I think I have addressed all your comments with that last push.

maxbrunsfeld commented:
Thanks @hendrikvanantwerpen!

hendrikvanantwerpen commented:
Thanks for accepting my changes, @maxbrunsfeld! I have one more question: We need to depend on these changes in a public library. Do you think it is possible to have a patch release with these changes relatively soon?
